modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/A11y.es.js

```diff
@@ -0,0 +1,272 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {ClayButtonWithIcon} from '@clayui/button';
+import ClayIcon from '@clayui/icon';
+import ClayPopover from '@clayui/popover';
+import observeRect from '@reach/observe-rect';
+import axe from 'axe-core';
+import React, {useEffect, useMemo, useRef, useState} from 'react';
+import ReactDOM from 'react-dom';
+
+import './A11y.css';
+
+function segmentAlertsByNode(alerts) {
+	const nodes = alerts.flatMap((alert) =>
+		alert.nodes.flatMap((node) => node.target)
+	);
+
+	return nodes.map((node) => {
+		const alertsByNode = alerts.filter((alert) =>
+			alert.nodes.some((n) => n.target.includes(node))
+		);
```

Comment:
This looks at least quadratic-ish (`filter` = one loop; `some` = another loop; `includes` = another loop). Can we fold this into the previous loops (the `flatMap` calls) in some way, seeing as you're already visiting every node there?

(I haven't taken the time to fully understand this, and I know you are calling this from `useMemo`, but that doesn't mean we shouldn't still try to make it more efficient.)

Reply:
@wincent yeah, I think, I can improve this a little bit and avoid going through other loops. The data is returned with all the rules violated and in each rule, there is a list of elements that are infringing. So it is simpler to try to take all the elements that are violating and add the rules that each one is violated.